Running bench/softmax.exs in EXLA on M1 occurs EXLA.Logger error

Hi,

I did the following steps on M1 and M1 Max Mac:

1. `git clone` the Nx repository;
2. `cd nx/exla`
3. `mix deps.get`
4. `mix run -r bench/softmax.exs`

Then, I met the following error:

```
** (Mix) Could not start application exla: EXLA.Application.start(:normal, []) returned an error: shutdown: failed to start child: EXLA.Logger
    ** (EXIT) an exception was raised:
        ** (UndefinedFunctionError) function EXLA.NIF.start_log_sink/1 is undefined (module EXLA.NIF is not available)
            (exla 0.1.0-dev) EXLA.NIF.start_log_sink(#PID<0.1989.0>)
            (exla 0.1.0-dev) lib/exla/logger.ex:12: EXLA.Logger.init/1
            (stdlib 3.17) gen_server.erl:423: :gen_server.init_it/2
            (stdlib 3.17) gen_server.erl:390: :gen_server.init_it/6
            (stdlib 3.17) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
```

I tested the following Elixir and Erlang environments and I met the above error on the both environments:

```
Erlang/OTP 24 [erts-12.2.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]

Elixir 1.13.3 (compiled with Erlang/OTP 24)
```

```
Erlang/OTP 25 [RELEASE CANDIDATE 1] [erts-13.0] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

Elixir 1.13.3 (compiled with Erlang/OTP 24)
```

No.

I got the following log:

```
% rm -rf /Users/zacky/Library/Caches/xla/0.3.0           
% rm -rf cache
% mix run -r bench/softmax.exs                

16:38:58.218 [info]  Found a matching archive (xla_extension-aarch64-darwin-cpu.tar.gz), going to download it
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   678  100   678    0     0   1864      0 --:--:-- --:--:-- --:--:--  1904
100 75.1M  100 75.1M    0     0  5924k      0  0:00:12  0:00:12 --:--:-- 8437k

16:39:11.231 [info]  Successfully downloaded the XLA archive
Unpacking /Users/zacky/Library/Caches/xla/0.3.0/cache/download/xla_extension-aarch64-darwin-cpu.tar.gz into /Users/zacky/github/nx/exla/cache
c++ -fPIC -I/Users/zacky/.asdf/installs/erlang/25.0-rc1/erts-13.0/include -Icache/xla_extension/include -O3 -Wall -Wno-sign-compare -Wno-unused-parameter -Wno-missing-field-initializers -Wno-comment -shared -std=c++14 c_src/exla/exla.cc c_src/exla/exla_nif_util.cc c_src/exla/exla_client.cc -o cache/libexla.so -Lcache/xla_extension/lib -lxla_extension -flat_namespace -undefined suppress
In file included from c_src/exla/exla.cc:4:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:28:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/heap_simulator.h:29:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/hlo_alias_analysis.h:25:
cache/xla_extension/include/tensorflow/compiler/xla/service/hlo_buffer.h:100:46: warning: 'color' is deprecated: Use Layout::memory_space instead. [-Wdeprecated-declarations]
    BufferValue::Color result = values()[0]->color();
                                             ^
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_value.h:111:3: note: 'color' has been explicitly marked deprecated here
  ABSL_DEPRECATED("Use Layout::memory_space instead.")
  ^
cache/xla_extension/include/absl/base/attributes.h:648:49: note: expanded from macro 'ABSL_DEPRECATED'
#define ABSL_DEPRECATED(message) __attribute__((deprecated(message)))
                                                ^
In file included from c_src/exla/exla.cc:4:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:28:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/heap_simulator.h:29:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/hlo_alias_analysis.h:25:
cache/xla_extension/include/tensorflow/compiler/xla/service/hlo_buffer.h:102:32: warning: 'color' is deprecated: Use Layout::memory_space instead. [-Wdeprecated-declarations]
      DCHECK_EQ(result, value->color());
                               ^
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_value.h:111:3: note: 'color' has been explicitly marked deprecated here
  ABSL_DEPRECATED("Use Layout::memory_space instead.")
  ^
cache/xla_extension/include/absl/base/attributes.h:648:49: note: expanded from macro 'ABSL_DEPRECATED'
#define ABSL_DEPRECATED(message) __attribute__((deprecated(message)))
                                                ^
In file included from c_src/exla/exla.cc:4:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:37:
cache/xla_extension/include/tensorflow/compiler/xla/service/memory_space_assignment.h:1000:1: warning: 'Options' defined as a struct here but previously declared as a class; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
struct Options {
^
cache/xla_extension/include/tensorflow/compiler/xla/service/memory_space_assignment.h:37:1: note: did you mean struct here?
class Options;
^~~~~
struct
In file included from c_src/exla/exla.cc:4:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:600:18: warning: 'set_color' is deprecated: Use Layout::memory_space instead. [-Wdeprecated-declarations]
          value->set_color(BufferValue::Color(
                 ^
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_value.h:118:3: note: 'set_color' has been explicitly marked deprecated here
  ABSL_DEPRECATED("Use Layout::memory_space instead.")
  ^
cache/xla_extension/include/absl/base/attributes.h:648:49: note: expanded from macro 'ABSL_DEPRECATED'
#define ABSL_DEPRECATED(message) __attribute__((deprecated(message)))
                                                ^
In file included from c_src/exla/exla.cc:4:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:603:18: warning: 'set_color' is deprecated: Use Layout::memory_space instead. [-Wdeprecated-declarations]
          value->set_color(BufferValue::Color(0));
                 ^
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_value.h:118:3: note: 'set_color' has been explicitly marked deprecated here
  ABSL_DEPRECATED("Use Layout::memory_space instead.")
  ^
cache/xla_extension/include/absl/base/attributes.h:648:49: note: expanded from macro 'ABSL_DEPRECATED'
#define ABSL_DEPRECATED(message) __attribute__((deprecated(message)))
                                                ^
5 warnings generated.
In file included from c_src/exla/exla_client.cc:1:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:28:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/heap_simulator.h:29:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/hlo_alias_analysis.h:25:
cache/xla_extension/include/tensorflow/compiler/xla/service/hlo_buffer.h:100:46: warning: 'color' is deprecated: Use Layout::memory_space instead. [-Wdeprecated-declarations]
    BufferValue::Color result = values()[0]->color();
                                             ^
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_value.h:111:3: note: 'color' has been explicitly marked deprecated here
  ABSL_DEPRECATED("Use Layout::memory_space instead.")
  ^
cache/xla_extension/include/absl/base/attributes.h:648:49: note: expanded from macro 'ABSL_DEPRECATED'
#define ABSL_DEPRECATED(message) __attribute__((deprecated(message)))
                                                ^
In file included from c_src/exla/exla_client.cc:1:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:28:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/heap_simulator.h:29:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/hlo_alias_analysis.h:25:
cache/xla_extension/include/tensorflow/compiler/xla/service/hlo_buffer.h:102:32: warning: 'color' is deprecated: Use Layout::memory_space instead. [-Wdeprecated-declarations]
      DCHECK_EQ(result, value->color());
                               ^
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_value.h:111:3: note: 'color' has been explicitly marked deprecated here
  ABSL_DEPRECATED("Use Layout::memory_space instead.")
  ^
cache/xla_extension/include/absl/base/attributes.h:648:49: note: expanded from macro 'ABSL_DEPRECATED'
#define ABSL_DEPRECATED(message) __attribute__((deprecated(message)))
                                                ^
In file included from c_src/exla/exla_client.cc:1:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:37:
cache/xla_extension/include/tensorflow/compiler/xla/service/memory_space_assignment.h:1000:1: warning: 'Options' defined as a struct here but previously declared as a class; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
struct Options {
^
cache/xla_extension/include/tensorflow/compiler/xla/service/memory_space_assignment.h:37:1: note: did you mean struct here?
class Options;
^~~~~
struct
In file included from c_src/exla/exla_client.cc:1:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:600:18: warning: 'set_color' is deprecated: Use Layout::memory_space instead. [-Wdeprecated-declarations]
          value->set_color(BufferValue::Color(
                 ^
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_value.h:118:3: note: 'set_color' has been explicitly marked deprecated here
  ABSL_DEPRECATED("Use Layout::memory_space instead.")
  ^
cache/xla_extension/include/absl/base/attributes.h:648:49: note: expanded from macro 'ABSL_DEPRECATED'
#define ABSL_DEPRECATED(message) __attribute__((deprecated(message)))
                                                ^
In file included from c_src/exla/exla_client.cc:1:
In file included from c_src/exla/exla_client.h:12:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/gpu_device.h:22:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.h:34:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/client/local_client.h:27:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/service/compiler.h:30:
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_assignment.h:603:18: warning: 'set_color' is deprecated: Use Layout::memory_space instead. [-Wdeprecated-declarations]
          value->set_color(BufferValue::Color(0));
                 ^
cache/xla_extension/include/tensorflow/compiler/xla/service/buffer_value.h:118:3: note: 'set_color' has been explicitly marked deprecated here
  ABSL_DEPRECATED("Use Layout::memory_space instead.")
  ^
cache/xla_extension/include/absl/base/attributes.h:648:49: note: expanded from macro 'ABSL_DEPRECATED'
#define ABSL_DEPRECATED(message) __attribute__((deprecated(message)))
                                                ^
5 warnings generated.
install_name_tool -change bazel-out/darwin_arm64-opt/bin/tensorflow/compiler/xla/extension/libxla_extension.so @loader_path/lib/libxla_extension.so -change bazel-out/darwin-opt/bin/tensorflow/compiler/xla/extension/libxla_extension.so @loader_path/lib/libxla_extension.so cache/libexla.so

16:39:49.333 [warning]  The on_load function for module Elixir.EXLA.NIF returned:
{:error,
 {:load_failed,
  'Failed to load NIF library: \'dlopen(/Users/zacky/github/nx/exla/_build/dev/lib/exla/priv/libexla.so, 0x0002): Library not loaded: @loader_path/lib/libxla_extension.so\n  Referenced from: /Users/zacky/github/nx/exla/cache/libexla.so\n  Reason: tried: \'/Users/zacky/github/nx/exla/cache/lib/libxla_extension.so\' (no such file), \'/usr/local/lib/libxla_extension.so\' (no such file), \'/usr/lib/libxla_extension.so\' (no such file)\''}}


16:39:49.334 [notice] domain=otp file=application_controller.erl line=2119  Application exla exited: EXLA.Application.start(:normal, []) returned an error: shutdown: failed to start child: EXLA.Logger
    ** (EXIT) an exception was raised:
        ** (UndefinedFunctionError) function EXLA.NIF.start_log_sink/1 is undefined (module EXLA.NIF is not available)
            (exla 0.1.0-dev) EXLA.NIF.start_log_sink(#PID<0.279.0>)
            (exla 0.1.0-dev) lib/exla/logger.ex:12: EXLA.Logger.init/1
            (stdlib 4.0) gen_server.erl:430: :gen_server.init_it/2
            (stdlib 4.0) gen_server.erl:397: :gen_server.init_it/6
            (stdlib 4.0) proc_lib.erl:226: :proc_lib.init_p_do_apply/3

16:39:49.360 [notice] domain=otp file=application_controller.erl line=2119  Application ex_doc exited: :stopped

16:39:49.360 [notice] domain=otp file=application_controller.erl line=2119  Application makeup_erlang exited: :stopped

16:39:49.361 [notice] domain=otp file=application_controller.erl line=2119  Application makeup_elixir exited: :stopped

16:39:49.361 [notice] domain=otp file=application_controller.erl line=2119  Application makeup exited: :stopped

16:39:49.361 [notice] domain=otp file=application_controller.erl line=2119  Application nimble_parsec exited: :stopped

16:39:49.361 [notice] domain=otp file=application_controller.erl line=2119  Application earmark_parser exited: :stopped

16:39:49.361 [notice] domain=otp file=application_controller.erl line=2119  Application eex exited: :stopped

16:39:49.361 [notice] domain=otp file=application_controller.erl line=2119  Application benchee exited: :stopped

16:39:49.361 [notice] domain=otp file=application_controller.erl line=2119  Application statistex exited: :stopped

16:39:49.361 [notice] domain=otp file=application_controller.erl line=2119  Application deep_merge exited: :stopped

16:39:49.361 [notice] domain=otp file=application_controller.erl line=2119  Application nx exited: :stopped
** (Mix) Could not start application exla: EXLA.Application.start(:normal, []) returned an error: shutdown: failed to start child: EXLA.Logger
    ** (EXIT) an exception was raised:
        ** (UndefinedFunctionError) function EXLA.NIF.start_log_sink/1 is undefined (module EXLA.NIF is not available)
            (exla 0.1.0-dev) EXLA.NIF.start_log_sink(#PID<0.279.0>)
            (exla 0.1.0-dev) lib/exla/logger.ex:12: EXLA.Logger.init/1
            (stdlib 4.0) gen_server.erl:430: :gen_server.init_it/2
            (stdlib 4.0) gen_server.erl:397: :gen_server.init_it/6
            (stdlib 4.0) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
```

--- a/exla/Makefile
+++ b/exla/Makefile
@@ -26,7 +26,7 @@ LDFLAGS = -L$(XLA_EXTENSION_LIB) -lxla_extension
 ifeq ($(shell uname -s), Darwin)
 	LDFLAGS += -flat_namespace -undefined suppress
 	POST_INSTALL = install_name_tool \
-		-change bazel-out/darwin_arm64-opt/bin/tensorflow/compiler/xla/extension/libxla_extension.so @loader_path/lib/libxla_extension.so \
+		-change bazel-out/darwin_arm64-opt/bin/tensorflow/compiler/xla/extension/libxla_extension.so @loader_path/xla_extension/lib/libxla_extension.so \
 		-change bazel-out/darwin-opt/bin/tensorflow/compiler/xla/extension/libxla_extension.so @loader_path/lib/libxla_extension.so \
 		$(EXLA_CACHE_SO)
 else

